cardano-simple/src/Cardano/Simple/PlutusLedgerApi/V1/Scripts.hs

```diff
@@ -78,6 +83,32 @@ mkStakeValidatorScriptPlutarch :: Config -> ClosedTerm a -> Either Text StakeVal
 mkStakeValidatorScriptPlutarch conf term =
   StakeValidator . Script . Plutarch.unScript <$> compile conf term
 
+mkPlutarchTypedScript :: Config -> ClosedTerm a -> Either Text (PlutarchTypedScript a)
+mkPlutarchTypedScript config term =
+  PlutarchTypedScript . Script . Plutarch.unScript <$> Plutarch.compile config term
+
+applyPlutarchTypedScript ::
+  Config ->
+  PlutarchTypedScript (a :--> b) ->
+  ClosedTerm a ->
+  Either Text (PlutarchTypedScript b)
+applyPlutarchTypedScript
+  config
+  (PlutarchTypedScript (Script (UPLC.Program _ v1 t1)))
+  term = do
+    (Plutarch.Script (UPLC.Program _ v2 t2)) <- Plutarch.compile config term
+
+    when (v1 GHC./= v2) $ throwError "Script versions differ"
+
+    pure $
+      PlutarchTypedScript $
+        Script $
+          UPLC.Program () v1 $
+            UPLC.Apply () t1 t2
+
+-- | 'PlutarchTypedScript' represents compiled a plutarch script while preserving type of the script
+newtype PlutarchTypedScript s = PlutarchTypedScript {getPlutarchTypedScript :: Script}
+
```

Comment:
Why say this is a Plutarch Script? It's a UPLC Script with a Plutarch type annotation right? I sense we could have all this Plutarch agnostic.

Reply:
Well, in principle you're right. For our needs we only needed plutarch annotations. That's why the only function that returns term of that type is `mkPlutarchTypedScript :: Config -> ClosedTerm a -> Either Text (PlutarchTypedScript a)`. Constructor of `PlutarchTypedScript` is exported. So the only way for the library user is to create a term either with `mkPlutarchTypedScript` or `applyPlutarchTypedScript`. If there is a need PlutusTx version of `mkPlutarchTypedScript` could be easily added. I don't do this right now because there is no business need for this in my project.